### PR TITLE
JW-296 Moved all initialization/deinitialization to `SxExtCreateSwitch` / `SxExtDeleteSwitch`

### PR DESCRIPTION
**Pull request by Attil**

### CODE REVIEW

File: windows/vrouter_mod.c [hunk 1]
```diff
@@ -215,43 +201,78 @@ SxExtCreateSwitch(
...
+    if (SxSwitchObject != NULL)
+        return NDIS_STATUS_FAILURE;
+
+    vr_num_cpus = KeQueryActiveProcessorCount(NULL);
+    if (!vr_num_cpus) {
+        DbgPrint("%s: Failed to get processor count\n", __func__);
+        return NDIS_STATUS_FAILURE;
+    }
+
+    SxSwitchObject = Switch;
```
Comment (m-kostrzewa) [inline windows/vrouter_mod.c]: SxSwitchObject is declared way before it is actually used. This ties in to my other comment about gotos.

File: windows/vrouter_mod.c [hunk 2]
```diff
@@ -21,6 +21,7 @@ ULONG  SxExtOidRequestId = 'RVCO';
 
 PSX_SWITCH_OBJECT SxSwitchObject = NULL;
 NDIS_HANDLE SxNBLPool = NULL;
+extern PDRIVER_OBJECT SxDriverObject;
```
Comment (sodar) [inline windows/vrouter_mod.c]: Maybe we should annotate `SxDriverObject` with a note that, SxBase exports this symbol.

File: windows/vrouter_mod.c [hunk 3]
```diff
@@ -215,43 +201,78 @@ SxExtCreateSwitch(
...
+cleanup_assoc:
+    vrouter_exit(false);
+cleanup_init:
+    vr_message_exit();
+cleanup_message:
+    DestroyDevice(SxDriverObject);
+cleanup_device:
+
+    SxSwitchObject = NULL;
+
```
Comment (m-kostrzewa) [inline windows/vrouter_mod.c]: Too many cleanup labels - this is very prone to error and an antipattern.

Consider something like this where applicable: http://stackoverflow.com/a/24215512. Calls to ExAllocatePoolWithTag and both NdisAllocateRWLock could be done this way. 

Also, try to refactoring a subroutine. Calls to vr_message_init, vrouter_init, vr_init_assoc could be grouped this way. You will need to create a cleanup subroutine for them, too.

Also, consider using a ```bool failed``` flag instead of so many gotos. A single ```cleanup``` label, or maybe 2-3 are fine, but 8 is way too many.

File: windows/vrouter_mod.c [hunk 4]
```diff
@@ -273,6 +296,8 @@ SxExtDeleteSwitch(
 
     vr_clean_assoc();
     vrouter_exit(false);
+    vr_message_exit();
```
Comment (sodar) [inline windows/vrouter_mod.c]: Analogically to @m-kostrzewa comment about initialization/goto labels in `SxExtCreateSwitch` - exit from vRouter and cleanup of Windows-related structures could be divided into multiple functions.

File: windows/vrouter_mod.c [hunk 5]
```diff
@@ -264,6 +285,8 @@ SxExtDeleteSwitch(
     DbgPrint("SxExtDeleteSwitch\r\n");
     UNREFERENCED_PARAMETER(Switch);
 
+    ASSERTMSG("Trying to delete another switch than currently active", Switch == SxSwitchObject);
```
Comment (sodar) [inline windows/vrouter_mod.c]: This check will not be effective in release build, thus an error could occur in production environment.

File: windows/vrouter_mod.c (hunk 1 above)
Comment (Attil) [inline windows/vrouter_mod.c]: Please note this is an assignment, not a declaration. Treat it like a `init_something`. `SxSwitchObject` is a global variable, which can be used by initialization functions used later.

File: windows/vrouter_mod.c (hunk 5 above)
Comment (Attil) [inline windows/vrouter_mod.c]: Could you please explain why?

Comment (sodar) [inline windows/vrouter_mod.c]: According to MSDN (https://msdn.microsoft.com/en-us/library/windows/hardware/ff542113(v=vs.85).aspx) this macro will not be included in binary on release builds. 

If a real life case when SxDeleteSwitch can be called on incorrect switch (when switch objects do not match) can occur, then macro will be ineffective in production. 

However, if this case can only be caused by programmer's error, then assert is ok.

Comment (sodar) [inline windows/vrouter_mod.c]: To clarify - can this assertion be violated in production in any way?

**Comment by Attil** [comment]
Bump

File: windows/vrouter_mod.c [hunk 6]
```diff
@@ -175,83 +184,173 @@ AddNicToArray(struct vr_switch_context* ctx, struct vr_nic* nic, NDIS_IF_COUNTED
...
+
 NDIS_STATUS
-SxExtCreateSwitch(
-    _In_ PSX_SWITCH_OBJECT Switch,
-    _Outptr_result_maybenull_ PNDIS_HANDLE *ExtensionContext
-)
+SxExtInitializeVRouter()
 {
-    DbgPrint("SxExtCreateSwitch\r\n");
+    BOOLEAN device = FALSE;
```
Comment (m-kostrzewa) [inline windows/vrouter_mod.c]: We had debate with @sodar and @renkam 

Ultimately we found a pretty nice book: https://books.google.pl/books?id=MTibAgAAQBAJ&pg=PT50&lpg=PT50&dq=kernel+goto+many+inits&source=bl&ots=ISMPwnVLxT&sig=ITso7AoC8-4cazRErrsV5gyZgvI&hl=en&sa=X&ved=0ahUKEwiE783e0YDTAhVTnRQKHSGRDaY4ChDoAQghMAE#v=onepage&q=kernel%20goto%20many%20inits&f=false

In the last example, the author proposes solution that uses global flags and a goto fail statement. We think that this might be the best option.

tl;dr:

```SxExtUninitializeVRouter(device, message, vrouter, assoc)``` should take no parameters. Instead, use global flags BOOLEAN message_status; etc. Then, use this structure:

``` 
message_status = vr_message_init()
if (!message_status) {
   goto cleanup;
}
...
cleanup:
   SxExtUninitializeVRouter()
```

Comment (m-kostrzewa) [inline windows/vrouter_mod.c]: BTW this is a pretty cool read, because the author compares "just gotos" with "single cleanup function" with this hybrid approach.

Comment (Attil) [inline windows/vrouter_mod.c]: Why do you think that approach is better?

Comment (Attil) [inline windows/vrouter_mod.c]: This presents some reasoning with which I agree:
http://wiki.c2.com/?GlobalVariablesAreBad

Comment (m-kostrzewa) [inline windows/vrouter_mod.c]: It looks cleaner. Functions with lots of parameters are also bad. And we mutate global state anyways. Otherwise, we would have code like:
```
Vrouter* myVrouter = vrouter_init();
```
Whether we like it or not, we keep messing with shared state, because we use callback architecture. 

Also, we call ```SxExtUninitializeVRouter``` with 4 ```TRUE``` params. If they were shared, we wouldn't have to make that assumption either.

We could consider keeping those state flags in Ctx, which are slightly less bad than global variables.

Comment (Attil) [inline windows/vrouter_mod.c]: Context is a compromise I can accept.

**Comment by Attil** [comment]
Bump

File: windows/vrouter_mod.c [hunk 7]
```diff
@@ -196,57 +196,55 @@ SxExtUninitialize(PDRIVER_OBJECT DriverObject)
...
-    BOOLEAN assoc = FALSE;
+    if (ctx->vrouter_up || ctx->device_up || ctx->message_up || ctx->assoc_up)
+        return NDIS_STATUS_FAILURE;
 
     if (!NT_SUCCESS(CreateDevice(SxDriverObject))) {
         goto cleanup;
     }
 
-    device = TRUE;
+    ctx->device_up = TRUE;
```
Comment (m-kostrzewa) [inline windows/vrouter_mod.c]: Why don't we try this instead?
```ctx->device_up = NT_SUCCESS(CreateDevice(SxDriverObject));
if(!ctx->device_up) {
	goto cleanup;
}

ctx->message_up = vr_message_init();
if(!ctx->message_up) {
	goto cleanup;
}

ctx->vrouter_up = vrouter_init();
if(!ctx->vrouter_up) {
	goto cleanup;
}

ctx->assoc_up = vr_init_assoc();
if(!ctx->assoc_up) {
	goto cleanup;
}
```

File: windows/vrouter_mod.c [hunk 8]
```diff
@@ -196,57 +196,55 @@ SxExtUninitialize(PDRIVER_OBJECT DriverObject)
...
-    if (message)
+    if (ctx->message_up)
         vr_message_exit();
 
-    if (vrouter)
+    if (ctx->vrouter_up)
         vrouter_exit(false);
 
-    if (assoc)
+    if (ctx->assoc_up)
```
Comment (m-kostrzewa) [inline windows/vrouter_mod.c]: I think if were to use braces (```{}```) for one line ```if``` statements everywhere else, then we should do it here, too.

Comment (Attil) [inline windows/vrouter_mod.c]: But we don't. Neither does dp-core (quickly looked at `vr_ip_mtrie.c:810`, `vr_datapath.c:42`.

File: windows/vrouter_mod.c (hunk 7 above)
Comment (Attil) [inline windows/vrouter_mod.c]: I'll go with `ctx->assoc_up = NT_SUCCESS(vr_init_assoc);`

Comment (Attil) [inline windows/vrouter_mod.c]: So it's a boolean and not a status.

Comment (Attil) [inline windows/vrouter_mod.c]: (I'll fix the lines above to conform to this standard)

File: windows/vrouter_mod.c (hunk 8 above)
Comment (Attil) [inline windows/vrouter_mod.c]: (I'll fix some lines to conform to this standard)

**Comment by Attil** [comment]
Bump

**Comment by sodar** [comment]
Is integration tests failure caused by the changes introduced by this PR or some error in test code in CI?

**Comment by Attil** [comment]
Janie przetestuj to proszę

I believe it's related to VM cleaning.

**Comment by Attil** [comment]
Janie, przetestuj proszę
